pkg/cafs/reader.go

```diff
@@ -159,6 +170,85 @@ func (r *chunkReader) WriteTo(writer io.Writer) (n int64, err error) {
 
 }
 
+func (r *chunkReader) calculateKeyAndOffset(off int64) (index int64, offset int64) {
+	index = off / int64(r.leafSize)
+	offset = off - (index * int64(r.leafSize))
+	return
+}
+
+func (r *chunkReader) ReadAt(p []byte, off int64) (n int, err error) {
+
+	// Calculate first key and offset.
+	index, offset := r.calculateKeyAndOffset(off)
+	if index >= int64(len(r.keys)) {
+		return 0, nil
+	}
+
+	var read int
+
+	seekHead := func() {
+		if index < int64(len(r.keys)-1) {
+			nextIndex := index + 1
+			if r.lru.Contains(r.keys[nextIndex].StringWithPrefix(r.prefix)) {
+				return
+			}
+			go func(i int64) {
+				k := r.keys[i]
+
+				rdr, e := r.fs.Get(context.Background(), k.StringWithPrefix(r.prefix))
+				if e != nil {
+					return
+				}
+
+				var b []byte
+				b, e = ioutil.ReadAll(rdr)
+				if e != nil {
+					return
+				}
+				r.lru.Add(k.StringWithPrefix(r.prefix), b)
+			}(nextIndex)
+		}
+	}
+
+	for {
+		// Fetch Blob
+		var buffer []byte
+		key := r.keys[index]
+		if r.lru.Contains(key.StringWithPrefix(r.prefix)) {
```

Comment:
on the first pass of the outer `for` loop, it's possible that this `if` block will need to account for `offset`